hack/generate_legacy_tests.py

```diff
@@ -0,0 +1,221 @@
+files were generated by kfctl as opposed to checked into kubeflow/manifests.
+
+Therefore to generate unittests we wanted to create kustomization.yaml files
+that are similar to those generated by kfctl. These kustomization.yaml files
+can then be used to generate the golden set of expected output by
+running kustomize build.
+
+Here's how this works.
+
+1. The script takes as input to the path to a ${KFAPP}/${KFDEF}.yaml
+   for which kfctl build was run.
+
+2. The script recourses over all the directories in ${KFAPP}/kustomize
+
+3. For each ${KFAPP}/kustomize/{APP}/kustomization.yaml the script
+   creates an a corresponding to test in
+   ${GIT_MANIFESTS}/tests/legacy/${APP}/kustomization.yaml
+
+This is primarily intended as onetime script. Once the kustomization.yaml
+files are generated and checked in there should be no reason to rerun it.
+
+Related issues:https://github.com/kubeflow/manifests/issues/1014
+"""
+import fire
+import logging
+import os
+import shutil
+import yaml
+
+# Which apps to skip. These are apps which had some edge case that it wasn't
+# worth dealing with.  It would just make more sense to generate
+# the tests manually
+APPS_TO_SKIP = ["mysql", "minio"]
+
+def build_configmap_patch(kustomize_dir):
+  kustomize_file = os.path.join(kustomize_dir, "kustomization.yaml")
+
+  with open(kustomize_file) as hf:
+    kustomize = yaml.load(hf)
+
+  params = {}
+  config_maps = []
+
+  for g in kustomize.get("configMapGenerator", []):
+    p_files = g.get("envs", [])
+
+    if "env" in g:
+      p_files.append(g["env"])
+
+    for p_file in p_files:
+      # If there is a params.env file add it as a config map patch
+      params_path = os.path.join(kustomize_dir, p_file)
+
+      if os.path.exists(params_path):
+        with open(params_path) as hf:
+          while True:
+            line = hf.readline()
+            if not line:
+              break
+            if not "=" in line:
+              continue
+            line = line.strip()
+            key, value = line.split("=", 1)
+            params[key] = value
+
+
+    # Generate a config map patch
+    patch = {
+      "apiVersion": "v1",
+      "kind": "ConfigMap",
+      "data": params,
+      "metadata": {
+        "name": g["name"],
+      }
+    }
+
+    config_maps.append(patch)
+
+  return config_maps
+
+def build_configmap_generators(kustomize_dir):
+  """If the file has a configmap generator we need to
+
+
+  Returns:
+   dict: config map name to list of files used for the configmap generator
+  """
+  kustomize_file = os.path.join(kustomize_dir, "kustomization.yaml")
+
+  with open(kustomize_file) as hf:
+    kustomize = yaml.load(hf)
+
+  generators = {}
+  for g in kustomize.get("configMapGenerator", []):
+    p_files = g.get("envs", [])
+
+    if "env" in g:
+      p_files.append(g["env"])
+
+    generators[g["name"]] = [os.path.join(kustomize_dir, f) for f in p_files]
+
+  return generators
+
+class GenerateLegacyTests:
```

Comment:
I should probably rename this to generate_legacy_kustomizations.py.
The tests are actually generated by generate_tests.py. In that directory you list the directories to search for kustomization.yaml to generate tests for.
https://github.com/kubeflow/manifests/blob/9cbb4d721d7b9e62da155956e3f2eae2a25e8b71/hack/generate_tests.py#L11

So if we have kustomizations that are worth testing we just include in the search directory.

The purpose of generate_legacy_tests.py is to account for the fact that right now we don't check our kustomization.yaml files into kubeflow/manifests and instead generate them via kfctl magic (#774).

So the way our tests are currently structured we generate a kustomization.yaml for kustomization.yaml we find in kubeflow/manifests. This kustomizations end up being trivial (i.e. not combining overlays or doing var substitutions). So we don't really end up testing that when all of the kustomizations are applied that we generate the expected YAML which is what we want to test.

Reply:
Thanks for the explanation.
If I'm understanding correctly:
1. Until the time stacks are implemented, we would be following this current approach of testing.
2. If we looking to generate a comprehensive test set we should look to use all supported kfdefs.
3. We may have to manually check-in kustomizations which are not referred to by any kfdef in the path `tests/legacy_kustomizations/`.

It seems the kustomizations for kfctl_k8s_istio.yaml or kfctl_istio_dex.yaml are yet to be generated. We should generate them with this PR.

It might be fine to overlook kustomizations affected by point 3.

Reply:
I understand that the conflicting bit here seems to be the kustomization.yaml file being generated by kfctl. We can call it the unit which we are testing here. The expectation from a developer should be to check in a kustomization.yaml for each 'unit' module.

Some documentation on the README.md file can help.

Reply:
Per comment below I don't think we should generate additional coverage in this PR. This PR is strictly increasing coverage. The priority is to increase converage to unblock the refactoring of the kustomize manifests for the V3 kustomize patterns.

So I think we should merge this PR and then any additional test coverage (e.g. for additional KFDefs) can happen in parallel.

> The earlier version of the unit test mechanism used to ensure that if make generate is run, the added kustomization's test files are automatically added to the test folder. This would ensure coverage or inform a developer to add tests.

This mechanism still exists. `make generate` will automatically generate tests for all paths listed in https://github.com/kubeflow/manifests/blob/9cbb4d721d7b9e62da155956e3f2eae2a25e8b71/hack/generate_tests.py#L11.

Right now the only search path is "legacy_kustomizations" because those are the only kustomization.yaml's that are worth testing. All the other kustomization.yaml files are trivial.

Think of it this way. imagine we create "kubeflow/manifests/examples" and in that directory we start definining kustomizations that we provide as examples to customers. So imagine we have a kustomization for katib that does something complicated; e.g use 5 different overlays, 6 different patches and we want to verify that it all generates the expected YAML.

At that point we would just add "kubeflow/manifests/examples" to search_dirs and then we would start autogenerating tests for those kustomization.yamls.

This should become more clear in subsequent PRs as we will start checking in complicated kustomization.yamls and creating tests for them by adding them to generate_tests.py